nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -1478,3 +1513,44 @@ enum TrimWhere {
     End,
     StartAndEnd,
 }
+
+enum NormalizeForm {
+    NFC,
+    NFD,
+    NFKC,
+    NFKD,
+}
+
+impl NormalizeForm {
+    fn as_str(&self) -> &'static str {
+        match self {
+            NormalizeForm::NFC => "NFC",
+            NormalizeForm::NFD => "NFD",
+            NormalizeForm::NFKC => "NFKC",
+            NormalizeForm::NFKD => "NFKD",
+        }
+    }
+}
+
+impl FromStr for NormalizeForm {
+    type Err = ();
+
+    fn from_str(input: &str) -> Result<NormalizeForm, Self::Err> {
+        match input {
+            "NFC" => Ok(NormalizeForm::NFC),
+            "NFD" => Ok(NormalizeForm::NFD),
+            "NFKC" => Ok(NormalizeForm::NFKC),
+            "NFKD" => Ok(NormalizeForm::NFKD),
+            _ => Err(()),
+        }
+    }
+}
+
+fn unicode_normalize(s: &str, f: NormalizeForm) -> std::string::String {
+    match f {
```

Comment:
**suggestion:** Let's use the unicode-normalization crate's `is_nfX_quick` functions to check (without allocations) if `s` is already normalized, and only if that returns `No` or `Maybe` do the `s.nfX().collect()` work. That way we can avoid allocating in the (probably?) common case of "this string is already normalized".

I don't know if ASCII is guaranteed to be already normalized, but if it is then we have another data point for heap-allocated strings that we can use to check the need to normalize, which is the UTF-8 => UTF-16 mapping table that heap strings carry.

Reply:
**issue:** You're now correctly checking for already-normalized strings but you're still making a clone of the string data. We don't want to do that: Please instead return eg. `None` from this function to signify that the original `s` parameter in the `normalize` function already contains a normalized function.

That way we actually avoid making a clone of the string data entirely.

Reply:
ohhh... got it, yes... that makes sense! I refactored the helper function, I think we could avoid the extra clone now